File: final_front_end_angular/src/app/body/service/CalculService.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_URL } from 'src/app/constants/app.constants';
import {Calcul} from "../api/calcul/Calcul";
import { CalculMatrix } from '../api/calcul/CalculMatrix';
import { CalculSummary } from '../api/calcul/CalculSummary';

@Injectable({
    providedIn: 'root',
})
export class CalculService {
    readonly API_URL = API_URL.DEV;
    readonly ENDPOINTS_CALCULS = '/calculs';

    constructor(private httpClient: HttpClient) { }

    //GET
    getAllCalculsByPatientId(
        patientId: number,
        organizationsId: number[],
        eyeSide: string,
    ): Observable<any[]> {
        let params = new HttpParams().set('eyeSide', eyeSide);
        return this.httpClient.get<any[]>(
            `${this.API_URL}${this.ENDPOINTS_CALCULS}/patients/${patientId}/organizations/${organizationsId.join(',')}`,
            { params }
        );
    }

    getAllCalculsByExamIdEyeSidePowerTrue(
        examId: number,
        organizationsId: number[],
        eyeSide: string,
    ): Observable<CalculSummary[]> {
        let params = new HttpParams().set('eyeSide', eyeSide);
        return this.httpClient.get<any[]>(
            `${this.API_URL}${this.ENDPOINTS_CALCULS}/exam/${examId}/organizations/${organizationsId.join(',')}/prec/list`,
            { params }
        );
    }

    getCalculById(
        id: number,
        organizationsId: number[]
    ): Observable<CalculMatrix> {
        return this.httpClient.get<CalculMatrix>(
            `${this.API_URL}${this.ENDPOINTS_CALCULS}/${id}/organizations/${organizationsId.join(',')}`
        );
    }

    //PUT
    selectPower(idCalcul: number, power: number, organizationsId: number[]) {
        let params = new HttpParams()
            .set('idCalcul', idCalcul)
            .append('power', power);
        return this.httpClient.put(
            `${this.API_URL}${this.ENDPOINTS_CALCULS}/organizationsId/${organizationsId.join(',')}/select-power`, params
        );
    }

    //POST
    injectionOnEachWebSite(calculs: Calcul[], organizationsId: number[]) {
        return this.httpClient.post(
            `${this.API_URL}${this.ENDPOINTS_CALCULS}/organizationsId/${organizationsId.join(',')}`, calculs
        );
    }

    //DELETE
}

File: final_back_end_spring_boot/src/main/java/com/jorami/eyeapp/service/CalculService.java
package com.jorami.eyeapp.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.jorami.eyeapp.model.Calcul;
import com.jorami.eyeapp.model.Enum;
import com.jorami.eyeapp.model.Lens;

import java.util.List;
import java.util.Map;

public interface CalculService {

    Calcul getCalculById(Long id, List<Long> organizationsId);

    List<Calcul> getAllCalculsByPatientIdAndEyeSide(Long patientId, Enum.EyeSide eyeSide, List<Long> organizationsId);

    List<Calcul> getAllCalculsByExamIdEyeSidePowerTrue(Long examId, Enum.EyeSide eyeSide, List<Long> organizationsId);

    void loopOnCalculs(List<Calcul> calculs, List<Long> organizationsId);

    void updateSelectedDiopters(Long idCalcul, Float power, List<Long> organizationsId);
}
